File: library/src/main/java/com/afollestad/aesthetic/ColorIsDarkState.java
package com.afollestad.aesthetic;

import static android.support.annotation.RestrictTo.Scope.LIBRARY_GROUP;

import android.support.annotation.ColorInt;
import android.support.annotation.RestrictTo;
import io.reactivex.functions.BiFunction;

/** @author Aidan Follestad (afollestad) */
@RestrictTo(LIBRARY_GROUP)
final class ColorIsDarkState {

  @ColorInt private final int color;
  private final boolean isDark;

  private ColorIsDarkState(@ColorInt int color, boolean isDark) {
    this.color = color;
    this.isDark = isDark;
  }

  static ColorIsDarkState create(@ColorInt int color, boolean isDark) {
    return new ColorIsDarkState(color, isDark);
  }

  static BiFunction<Integer, Boolean, ColorIsDarkState> creator() {
    return new BiFunction<Integer, Boolean, ColorIsDarkState>() {
      @Override
      public ColorIsDarkState apply(Integer integer, Boolean aBoolean) {
        return ColorIsDarkState.create(integer, aBoolean);
      }
    };
  }

  @ColorInt
  int color() {
    return color;
  }

  boolean isDark() {
    return isDark;
  }
}

File: library/src/main/java/com/afollestad/aesthetic/ColorIsDarkState.kt
package com.afollestad.aesthetic

import android.support.annotation.ColorInt
import io.reactivex.functions.BiFunction

/** @author Aidan Follestad (afollestad)
 */
internal data class ColorIsDarkState(
  @field:ColorInt val color: Int,
  val isDark: Boolean
) {
  companion object {
    fun creator(): BiFunction<Int, Boolean, ColorIsDarkState> {
      return BiFunction { color, isDark -> ColorIsDarkState(color, isDark) }
    }
  }
}
